File: pkg/incidents/incidents.go
// Package incidents implements incident lifecycle management — shared by
// alert-service (which opens incidents automatically from new alerts) and
// the api service (which exposes manual transitions/assignment to a human
// operator). Extracted from alert-service once a second caller needed the
// exact same state machine and persistence, rather than duplicating it.
package incidents

import (
	"context"
	"errors"
	"fmt"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// AlertRef is the minimal alert data needed to open or attach an incident —
// deliberately independent of alert-service's own Alert type so this
// package has no dependency on it.
type AlertRef struct {
	ID             string
	OrganizationID string
	Severity       string
	FactoryID      string
	MachineID      string
	DeviceID       string
	SensorID       string
	Title          string
	Description    string
}

type Incident struct {
	ID              string
	OrganizationID  string
	AlertID         string
	FactoryID       string
	MachineID       string
	DeviceID        string
	SensorID        string
	Severity        string
	Status          string
	Title           string
	Description     string
	AssignedTo      string
	ResolutionNotes string
	OpenedAt        time.Time
	ResolvedAt      *time.Time
	ClosedAt        *time.Time
}

type Event struct {
	ID          string
	IncidentID  string
	EventType   string
	ActorUserID *string
	OldValue    string
	NewValue    string
	Note        string
	CreatedAt   time.Time
}

// Publisher lets Store announce lifecycle changes on the `incidents` Kafka
// topic. Optional — a nil Publisher (or NewStore called without one) simply
// skips publishing, since not every caller needs it.
type Publisher interface {
	PublishIncidentEvent(ctx context.Context, eventType string, inc Incident) error
}

// validTransitions defines the incident lifecycle state machine. RESOLVED
// can move back to INVESTIGATING (a recurrence) but CLOSED is terminal —
// once closed, a recurrence opens a new incident instead of reanimating an
// old one, keeping the audit trail honest about when the problem was first
// actually seen.
var validTransitions = map[string][]string{
	"OPEN":          {"ACKNOWLEDGED", "INVESTIGATING", "RESOLVED"},
	"ACKNOWLEDGED":  {"INVESTIGATING", "RESOLVED"},
	"INVESTIGATING": {"RESOLVED"},
	"RESOLVED":      {"CLOSED", "INVESTIGATING"},
	"CLOSED":        {},
}

func IsValidTransition(from, to string) bool {
	for _, allowed := range validTransitions[from] {
		if allowed == to {
			return true
		}
	}
	return false
}

type Store struct {
	pool      *pgxpool.Pool
	publisher Publisher
}

func NewStore(pool *pgxpool.Pool, publisher Publisher) *Store {
	return &Store{pool: pool, publisher: publisher}
}

// OpenOrAttach implements "alerts can result in incidents, but don't create
// unlimited incidents from repeated alerts": it reuses any incident already
// active for this machine (enforced by the partial unique index from
// Phase 2, so this is race-safe, not just an application-level check)
// rather than opening a second one, logging the new alert's arrival as an
// ALERT_ATTACHED audit event instead.
func (s *Store) OpenOrAttach(ctx context.Context, alert AlertRef) (incidentID string, created bool, err error) {
	var existingID string
	err = s.pool.QueryRow(ctx,
		`SELECT id FROM incidents WHERE machine_id = $1 AND status IN ('OPEN', 'ACKNOWLEDGED', 'INVESTIGATING')`,
		alert.MachineID,
	).Scan(&existingID)

	if err == nil {
		if attachErr := s.logEvent(ctx, existingID, "ALERT_ATTACHED", nil, "", "", fmt.Sprintf("alert %s (%s) attached: %s", alert.ID, alert.Severity, alert.Title)); attachErr != nil {
			return "", false, attachErr
		}
		return existingID, false, nil
	}
	if !errors.Is(err, pgx.ErrNoRows) {
		return "", false, fmt.Errorf("check existing incident: %w", err)
	}

	var id string
	err = s.pool.QueryRow(ctx,
		`INSERT INTO incidents (organization_id, alert_id, factory_id, machine_id, device_id, sensor_id, severity, status, title, description)
		 VALUES ($1, $2, $3, $4, $5, $6, $7, 'OPEN', $8, $9)
		 ON CONFLICT (machine_id) WHERE status IN ('OPEN', 'ACKNOWLEDGED', 'INVESTIGATING') DO NOTHING
		 RETURNING id`,
		alert.OrganizationID, alert.ID, nullIfEmpty(alert.FactoryID), nullIfEmpty(alert.MachineID), nullIfEmpty(alert.DeviceID), nullIfEmpty(alert.SensorID),
		alert.Severity, alert.Title, alert.Description,
	).Scan(&id)

	if errors.Is(err, pgx.ErrNoRows) {
		if selErr := s.pool.QueryRow(ctx,
			`SELECT id FROM incidents WHERE machine_id = $1 AND status IN ('OPEN', 'ACKNOWLEDGED', 'INVESTIGATING')`,
			alert.MachineID,
		).Scan(&existingID); selErr != nil {
			return "", false, fmt.Errorf("resolve incident creation race: %w", selErr)
		}
		return existingID, false, nil
	}
	if err != nil {
		return "", false, fmt.Errorf("insert incident: %w", err)
	}

	if logErr := s.logEvent(ctx, id, "STATUS_CHANGE", nil, "", "OPEN", "incident opened from alert "+alert.ID); logErr != nil {
		return "", false, logErr
	}
	s.publish(ctx, "CREATED", id)
	return id, true, nil
}

// Transition moves an incident to a new status if the move is valid,
// recording the change in incident_events. actorUserID is nil for
// system-initiated transitions (e.g. none currently) and set for
// human-initiated ones once the caller has a JWT to identify who acted.
func (s *Store) Transition(ctx context.Context, incidentID, newStatus string, actorUserID *string, note string) error {
	var currentStatus string
	if err := s.pool.QueryRow(ctx, `SELECT status FROM incidents WHERE id = $1`, incidentID).Scan(&currentStatus); err != nil {
		return fmt.Errorf("load incident %s: %w", incidentID, err)
	}
	if !IsValidTransition(currentStatus, newStatus) {
		return fmt.Errorf("invalid incident transition %s -> %s", currentStatus, newStatus)
	}

	var extra string
	switch newStatus {
	case "RESOLVED":
		extra = `, resolved_at = now()`
	case "CLOSED":
		extra = `, closed_at = now()`
	}

	_, err := s.pool.Exec(ctx,
		fmt.Sprintf(`UPDATE incidents SET status = $1, updated_at = now()%s WHERE id = $2`, extra),
		newStatus, incidentID,
	)
	if err != nil {
		return fmt.Errorf("update incident %s status: %w", incidentID, err)
	}

	if err := s.logEvent(ctx, incidentID, "STATUS_CHANGE", actorUserID, currentStatus, newStatus, note); err != nil {
		return err
	}
	s.publish(ctx, "STATUS_CHANGED", incidentID)
	return nil
}

func (s *Store) Assign(ctx context.Context, incidentID, userID string, actorUserID *string) error {
	_, err := s.pool.Exec(ctx, `UPDATE incidents SET assigned_to = $1, updated_at = now() WHERE id = $2`, userID, incidentID)
	if err != nil {
		return fmt.Errorf("assign incident %s: %w", incidentID, err)
	}
	if err := s.logEvent(ctx, incidentID, "ASSIGNMENT", actorUserID, "", userID, "assigned to technician "+userID); err != nil {
		return err
	}
	s.publish(ctx, "ASSIGNED", incidentID)
	return nil
}

func (s *Store) Resolve(ctx context.Context, incidentID, resolutionNotes string, actorUserID *string) error {
	if _, err := s.pool.Exec(ctx, `UPDATE incidents SET resolution_notes = $1 WHERE id = $2`, resolutionNotes, incidentID); err != nil {
		return fmt.Errorf("set resolution notes for incident %s: %w", incidentID, err)
	}
	return s.Transition(ctx, incidentID, "RESOLVED", actorUserID, resolutionNotes)
}

// Get fetches an incident scoped to orgID — a resource that doesn't belong
// to that organization returns pgx.ErrNoRows, indistinguishable from "does
// not exist," which is the correct tenant-isolation behavior (never reveal
// that a resource exists in another organization).
func (s *Store) Get(ctx context.Context, orgID, incidentID string) (*Incident, error) {
	var inc Incident
	err := s.pool.QueryRow(ctx, `
		SELECT id, organization_id, COALESCE(alert_id::text,''), COALESCE(factory_id::text,''),
		       COALESCE(machine_id::text,''), COALESCE(device_id::text,''), COALESCE(sensor_id::text,''),
		       severity, status, title, description, COALESCE(assigned_to::text,''), COALESCE(resolution_notes,''),
		       opened_at, resolved_at, closed_at
		FROM incidents WHERE id = $1 AND organization_id = $2
	`, incidentID, orgID).Scan(
		&inc.ID, &inc.OrganizationID, &inc.AlertID, &inc.FactoryID, &inc.MachineID, &inc.DeviceID, &inc.SensorID,
		&inc.Severity, &inc.Status, &inc.Title, &inc.Description, &inc.AssignedTo, &inc.ResolutionNotes,
		&inc.OpenedAt, &inc.ResolvedAt, &inc.ClosedAt,
	)
	if err != nil {
		return nil, err
	}
	return &inc, nil
}

// List returns incidents for orgID, optionally filtered by status, newest
// first, with simple offset pagination.
func (s *Store) List(ctx context.Context, orgID, statusFilter string, limit, offset int) ([]Incident, error) {
	query := `
		SELECT id, organization_id, COALESCE(alert_id::text,''), COALESCE(factory_id::text,''),
		       COALESCE(machine_id::text,''), COALESCE(device_id::text,''), COALESCE(sensor_id::text,''),
		       severity, status, title, description, COALESCE(assigned_to::text,''), COALESCE(resolution_notes,''),
		       opened_at, resolved_at, closed_at
		FROM incidents WHERE organization_id = $1`
	args := []any{orgID}
	if statusFilter != "" {
		query += " AND status = $2"
		args = append(args, statusFilter)
	}
	query += fmt.Sprintf(" ORDER BY opened_at DESC LIMIT %d OFFSET %d", limit, offset)

	rows, err := s.pool.Query(ctx, query, args...)
	if err != nil {
		return nil, fmt.Errorf("list incidents: %w", err)
	}
	defer rows.Close()

	var out []Incident
	for rows.Next() {
		var inc Incident
		if err := rows.Scan(
			&inc.ID, &inc.OrganizationID, &inc.AlertID, &inc.FactoryID, &inc.MachineID, &inc.DeviceID, &inc.SensorID,
			&inc.Severity, &inc.Status, &inc.Title, &inc.Description, &inc.AssignedTo, &inc.ResolutionNotes,
			&inc.OpenedAt, &inc.ResolvedAt, &inc.ClosedAt,
		); err != nil {
			return nil, fmt.Errorf("scan incident row: %w", err)
		}
		out = append(out, inc)
	}
	return out, rows.Err()
}

func (s *Store) ListEvents(ctx context.Context, incidentID string) ([]Event, error) {
	rows, err := s.pool.Query(ctx, `
		SELECT id, incident_id, event_type, actor_user_id, COALESCE(old_value,''), COALESCE(new_value,''), COALESCE(note,''), created_at
		FROM incident_events WHERE incident_id = $1 ORDER BY created_at ASC
	`, incidentID)
	if err != nil {
		return nil, fmt.Errorf("list incident events: %w", err)
	}
	defer rows.Close()

	var out []Event
	for rows.Next() {
		var e Event
		if err := rows.Scan(&e.ID, &e.IncidentID, &e.EventType, &e.ActorUserID, &e.OldValue, &e.NewValue, &e.Note, &e.CreatedAt); err != nil {
			return nil, fmt.Errorf("scan incident event: %w", err)
		}
		out = append(out, e)
	}
	return out, rows.Err()
}

func (s *Store) logEvent(ctx context.Context, incidentID, eventType string, actorUserID *string, oldValue, newValue, note string) error {
	_, err := s.pool.Exec(ctx,
		`INSERT INTO incident_events (incident_id, event_type, actor_user_id, old_value, new_value, note)
		 VALUES ($1, $2, $3, $4, $5, $6)`,
		incidentID, eventType, actorUserID, nullIfEmpty(oldValue), nullIfEmpty(newValue), note,
	)
	if err != nil {
		return fmt.Errorf("log incident event for %s: %w", incidentID, err)
	}
	return nil
}

// publish is best-effort: a Kafka outage must never fail the underlying
// Postgres mutation that already succeeded, so errors here are swallowed.
func (s *Store) publish(ctx context.Context, eventType, incidentID string) {
	if s.publisher == nil {
		return
	}
	full, err := s.fetchByID(ctx, incidentID)
	if err != nil {
		return
	}
	_ = s.publisher.PublishIncidentEvent(ctx, eventType, *full)
}

func (s *Store) fetchByID(ctx context.Context, incidentID string) (*Incident, error) {
	var inc Incident
	err := s.pool.QueryRow(ctx, `
		SELECT id, organization_id, COALESCE(alert_id::text,''), COALESCE(factory_id::text,''),
		       COALESCE(machine_id::text,''), COALESCE(device_id::text,''), COALESCE(sensor_id::text,''),
		       severity, status, title, description, COALESCE(assigned_to::text,''), COALESCE(resolution_notes,''),
		       opened_at, resolved_at, closed_at
		FROM incidents WHERE id = $1
	`, incidentID).Scan(
		&inc.ID, &inc.OrganizationID, &inc.AlertID, &inc.FactoryID, &inc.MachineID, &inc.DeviceID, &inc.SensorID,
		&inc.Severity, &inc.Status, &inc.Title, &inc.Description, &inc.AssignedTo, &inc.ResolutionNotes,
		&inc.OpenedAt, &inc.ResolvedAt, &inc.ClosedAt,
	)
	if err != nil {
		return nil, err
	}
	return &inc, nil
}

func nullIfEmpty(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

File: shared/incidents.py
"""Incident lifecycle management, the Python port of pkg/incidents.

Shared by alert-service (which opens incidents automatically from new
alerts) and the api service (which will expose manual transitions/
assignment to a human operator once it's ported). Extracted into shared/
for the same reason the Go version was pulled out of alert-service into
its own package once a second caller needed the exact same state machine
and persistence, rather than duplicating it.
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime
from typing import Protocol

from psycopg_pool import ConnectionPool


@dataclass
class AlertRef:
    """The minimal alert data needed to open or attach an incident —
    deliberately independent of alert-service's own Alert type so this
    module has no dependency on it."""

    id: str
    organization_id: str
    severity: str
    factory_id: str = ""
    machine_id: str = ""
    device_id: str = ""
    sensor_id: str = ""
    title: str = ""
    description: str = ""


@dataclass
class Incident:
    id: str
    organization_id: str
    alert_id: str
    factory_id: str
    machine_id: str
    device_id: str
    sensor_id: str
    severity: str
    status: str
    title: str
    description: str
    assigned_to: str
    resolution_notes: str
    opened_at: datetime
    resolved_at: datetime | None
    closed_at: datetime | None


@dataclass
class Event:
    id: str
    incident_id: str
    event_type: str
    actor_user_id: str | None
    old_value: str
    new_value: str
    note: str
    created_at: datetime


class Publisher(Protocol):
    """Lets Store announce lifecycle changes on the `incidents` Kafka
    topic. Optional — a Store built without one simply skips publishing,
    since not every caller needs it."""

    def publish_incident_event(self, event_type: str, inc: Incident) -> None: ...


# The incident lifecycle state machine. RESOLVED can move back to
# INVESTIGATING (a recurrence) but CLOSED is terminal — once closed, a
# recurrence opens a new incident instead of reanimating an old one,
# keeping the audit trail honest about when the problem was first actually
# seen.
_VALID_TRANSITIONS: dict[str, list[str]] = {
    "OPEN": ["ACKNOWLEDGED", "INVESTIGATING", "RESOLVED"],
    "ACKNOWLEDGED": ["INVESTIGATING", "RESOLVED"],
    "INVESTIGATING": ["RESOLVED"],
    "RESOLVED": ["CLOSED", "INVESTIGATING"],
    "CLOSED": [],
}


def is_valid_transition(from_status: str, to_status: str) -> bool:
    return to_status in _VALID_TRANSITIONS.get(from_status, [])


_INCIDENT_COLUMNS = """
    id, organization_id, COALESCE(alert_id::text,''), COALESCE(factory_id::text,''),
    COALESCE(machine_id::text,''), COALESCE(device_id::text,''), COALESCE(sensor_id::text,''),
    severity, status, title, description, COALESCE(assigned_to::text,''), COALESCE(resolution_notes,''),
    opened_at, resolved_at, closed_at
"""


def _row_to_incident(row) -> Incident:
    return Incident(
        # id/organization_id are plain uuid columns in this query (unlike
        # the others below, which are ::text-cast in SQL) -- cast here too.
        id=str(row[0]), organization_id=str(row[1]), alert_id=row[2], factory_id=row[3],
        machine_id=row[4], device_id=row[5], sensor_id=row[6],
        severity=row[7], status=row[8], title=row[9], description=row[10],
        assigned_to=row[11], resolution_notes=row[12],
        opened_at=row[13], resolved_at=row[14], closed_at=row[15],
    )


class Store:
    def __init__(self, pool: ConnectionPool, publisher: Publisher | None = None):
        self._pool = pool
        self._publisher = publisher

    def open_or_attach(self, alert: AlertRef) -> tuple[str, bool]:
        """Implements "alerts can result in incidents, but don't create
        unlimited incidents from repeated alerts": reuses any incident
        already active for this machine (enforced by a partial unique
        index, so this is race-safe, not just an application-level check)
        rather than opening a second one, logging the new alert's arrival
        as an ALERT_ATTACHED audit event instead. Returns (incident_id,
        created)."""
        with self._pool.connection() as conn:
            row = conn.execute(
                "SELECT id FROM incidents WHERE machine_id = %s AND status IN ('OPEN', 'ACKNOWLEDGED', 'INVESTIGATING')",
                (alert.machine_id,),
            ).fetchone()

            if row is not None:
                # id/uuid columns come back as uuid.UUID, not str -- cast
                # explicitly everywhere one is returned from a query.
                existing_id = str(row[0])
                self._log_event(
                    conn, existing_id, "ALERT_ATTACHED", None, "", "",
                    f"alert {alert.id} ({alert.severity}) attached: {alert.title}",
                )
                return existing_id, False

            row = conn.execute(
                """
                INSERT INTO incidents (organization_id, alert_id, factory_id, machine_id, device_id, sensor_id, severity, status, title, description)
                VALUES (%s, %s, %s, %s, %s, %s, %s, 'OPEN', %s, %s)
                ON CONFLICT (machine_id) WHERE status IN ('OPEN', 'ACKNOWLEDGED', 'INVESTIGATING') DO NOTHING
                RETURNING id
                """,
                (
                    alert.organization_id, alert.id, _null_if_empty(alert.factory_id), _null_if_empty(alert.machine_id),
                    _null_if_empty(alert.device_id), _null_if_empty(alert.sensor_id), alert.severity, alert.title, alert.description,
                ),
            ).fetchone()

            if row is None:
                # Lost a race with a concurrent insert for the same open incident.
                existing = conn.execute(
                    "SELECT id FROM incidents WHERE machine_id = %s AND status IN ('OPEN', 'ACKNOWLEDGED', 'INVESTIGATING')",
                    (alert.machine_id,),
                ).fetchone()
                if existing is None:
                    raise RuntimeError(f"resolve incident creation race for machine {alert.machine_id}: no incident found")
                return str(existing[0]), False

            incident_id = str(row[0])
            self._log_event(conn, incident_id, "STATUS_CHANGE", None, "", "OPEN", f"incident opened from alert {alert.id}")
        self._publish("CREATED", incident_id)
        return incident_id, True

    def transition(self, incident_id: str, new_status: str, actor_user_id: str | None, note: str) -> None:
        """Moves an incident to a new status if the move is valid,
        recording the change in incident_events. actor_user_id is None for
        system-initiated transitions and set for human-initiated ones once
        the caller has identified who acted."""
        with self._pool.connection() as conn:
            row = conn.execute("SELECT status FROM incidents WHERE id = %s", (incident_id,)).fetchone()
            if row is None:
                raise RuntimeError(f"load incident {incident_id}: not found")
            current_status = row[0]
            if not is_valid_transition(current_status, new_status):
                raise ValueError(f"invalid incident transition {current_status} -> {new_status}")

            extra = ""
            if new_status == "RESOLVED":
                extra = ", resolved_at = now()"
            elif new_status == "CLOSED":
                extra = ", closed_at = now()"

            conn.execute(f"UPDATE incidents SET status = %s, updated_at = now(){extra} WHERE id = %s", (new_status, incident_id))
            self._log_event(conn, incident_id, "STATUS_CHANGE", actor_user_id, current_status, new_status, note)
        self._publish("STATUS_CHANGED", incident_id)

    def assign(self, incident_id: str, user_id: str, actor_user_id: str | None) -> None:
        with self._pool.connection() as conn:
            conn.execute("UPDATE incidents SET assigned_to = %s, updated_at = now() WHERE id = %s", (user_id, incident_id))
            self._log_event(conn, incident_id, "ASSIGNMENT", actor_user_id, "", user_id, f"assigned to technician {user_id}")
        self._publish("ASSIGNED", incident_id)

    def resolve(self, incident_id: str, resolution_notes: str, actor_user_id: str | None) -> None:
        with self._pool.connection() as conn:
            conn.execute("UPDATE incidents SET resolution_notes = %s WHERE id = %s", (resolution_notes, incident_id))
        self.transition(incident_id, "RESOLVED", actor_user_id, resolution_notes)

    def get(self, org_id: str, incident_id: str) -> Incident | None:
        """Fetches an incident scoped to org_id — a resource that doesn't
        belong to that organization returns None, indistinguishable from
        "does not exist," which is the correct tenant-isolation behavior
        (never reveal that a resource exists in another organization)."""
        with self._pool.connection() as conn:
            row = conn.execute(
                f"SELECT {_INCIDENT_COLUMNS} FROM incidents WHERE id = %s AND organization_id = %s",
                (incident_id, org_id),
            ).fetchone()
        return _row_to_incident(row) if row is not None else None

    def list(self, org_id: str, status_filter: str, limit: int, offset: int) -> list[Incident]:
        """Returns incidents for org_id, optionally filtered by status,
        newest first, with simple offset pagination."""
        query = f"SELECT {_INCIDENT_COLUMNS} FROM incidents WHERE organization_id = %s"
        params: list = [org_id]
        if status_filter:
            query += " AND status = %s"
            params.append(status_filter)
        query += " ORDER BY opened_at DESC LIMIT %s OFFSET %s"
        params.extend([limit, offset])

        with self._pool.connection() as conn:
            rows = conn.execute(query, params).fetchall()
        return [_row_to_incident(row) for row in rows]

    def list_events(self, incident_id: str) -> list[Event]:
        with self._pool.connection() as conn:
            rows = conn.execute(
                """
                SELECT id, incident_id, event_type, actor_user_id, COALESCE(old_value,''), COALESCE(new_value,''), COALESCE(note,''), created_at
                FROM incident_events WHERE incident_id = %s ORDER BY created_at ASC
                """,
                (incident_id,),
            ).fetchall()
        return [
            Event(
                id=str(r[0]), incident_id=str(r[1]), event_type=r[2],
                actor_user_id=str(r[3]) if r[3] is not None else None,
                old_value=r[4], new_value=r[5], note=r[6], created_at=r[7],
            )
            for r in rows
        ]

    def _log_event(self, conn, incident_id: str, event_type: str, actor_user_id: str | None, old_value: str, new_value: str, note: str) -> None:
        conn.execute(
            """
            INSERT INTO incident_events (incident_id, event_type, actor_user_id, old_value, new_value, note)
            VALUES (%s, %s, %s, %s, %s, %s)
            """,
            (incident_id, event_type, actor_user_id, _null_if_empty(old_value), _null_if_empty(new_value), note),
        )

    def _publish(self, event_type: str, incident_id: str) -> None:
        """Best-effort: a Kafka outage must never fail the underlying
        Postgres mutation that already succeeded, so errors here are
        swallowed."""
        if self._publisher is None:
            return
        full = self._fetch_by_id(incident_id)
        if full is None:
            return
        try:
            self._publisher.publish_incident_event(event_type, full)
        except Exception:  # noqa: BLE001
            pass

    def _fetch_by_id(self, incident_id: str) -> Incident | None:
        with self._pool.connection() as conn:
            row = conn.execute(f"SELECT {_INCIDENT_COLUMNS} FROM incidents WHERE id = %s", (incident_id,)).fetchone()
        return _row_to_incident(row) if row is not None else None


def _null_if_empty(s: str) -> str | None:
    return s if s else None
